### PR TITLE
gh-117431: Adapt bytes and bytearray .find() and friends to Argument Clinic

### PR DESCRIPTION
**Pull request by erlend-aasland**
This change gives a significant speedup, as the METH_FASTCALL calling convention is used.

* Issue: gh-117431

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Note that the diff is not as frightening as GitHub wants it; subtract the generated files, and the diff is within ~200 lines of code, most of it being clinic input blocks.

Moreover, we want to update either this PR or #117468 to remove the auxiliary stringlib functions, depending on which PR lands first.

**Comment by erlend-aasland** [comment]
@methane, I intend to merge this in time for the beta in May. Aiming to land it next week.

File: Objects/bytearrayobject.c [hunk 1]
```diff
@@ -1121,16 +1121,44 @@ bytearray_dealloc(PyByteArrayObject *self)
...
+bytearray.find
+
+    sub: object
+    start: slice_index(accept={int, NoneType}, c_default='0') = None
+         Optional start position. Default: start of the bytes.
+    end: slice_index(accept={int, NoneType}, c_default='PY_SSIZE_T_MAX') = None
+         Optional stop position. Default: end of the bytes.
+    /
+
+Return the lowest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start,end].
```
Comment (methane) [inline Objects/bytearrayobject.c]:
```suggestion
Return the lowest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start:end].
```

File: Objects/bytearrayobject.c [hunk 2]
```diff
@@ -1162,22 +1190,55 @@ bytearray_copy_impl(PyByteArrayObject *self)
                                          PyByteArray_GET_SIZE(self));
 }
 
+/*[clinic input]
+bytearray.index = bytearray.find
+
+Return the lowest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start,end].
```
Comment (methane) [inline Objects/bytearrayobject.c]:
```suggestion
Return the lowest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start:end].
```

File: Objects/bytearrayobject.c [hunk 3]
```diff
@@ -1162,22 +1190,55 @@ bytearray_copy_impl(PyByteArrayObject *self)
...
 {
-    return _Py_bytes_index(PyByteArray_AS_STRING(self), PyByteArray_GET_SIZE(self), args);
+    return _Py_bytes_index(PyByteArray_AS_STRING(self), PyByteArray_GET_SIZE(self),
+                           sub, start, end);
 }
 
+/*[clinic input]
+bytearray.rfind = bytearray.find
+
+Return the highest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start,end].
```
Comment (methane) [inline Objects/bytearrayobject.c]:
```suggestion
Return the highest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start:end].
```

File: Objects/bytearrayobject.c [hunk 4]
```diff
@@ -1162,22 +1190,55 @@ bytearray_copy_impl(PyByteArrayObject *self)
...
 {
-    return _Py_bytes_rfind(PyByteArray_AS_STRING(self), PyByteArray_GET_SIZE(self), args);
+    return _Py_bytes_rfind(PyByteArray_AS_STRING(self), PyByteArray_GET_SIZE(self),
+                           sub, start, end);
 }
 
+/*[clinic input]
+bytearray.rindex = bytearray.find
+
+Return the highest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start,end].
```
Comment (methane) [inline Objects/bytearrayobject.c]:
```suggestion
Return the highest index in B where subsection 'sub' is found, such that 'sub' is contained within B[start:end].
```

**Review by methane — COMMENTED** [review]
Although old docstring used [start,end], I want to fix them to [start:end].

**Comment by erlend-aasland** [comment]
Thanks for the review, Inada-san!

**Pull request merged**